NPE in CachingTaskDependencyResolveContext.add

<!--- 
Please use our bug report template to report problems with something that has never worked.  Regressions reports are greatly appreciated during our RC phase and before a final release.
-->
our build that works fine with 5.5.rcx and 5.4.1 fails with 5.6 5.6-20190611000032+0000
<!--- Provide a brief summary of the issue in the title above -->

### Expected Behavior
<!--- Tell us what should happen -->
Build works fine
### Current Behavior
<!--- Tell us what happens instead of the expected behavior -->
Build Fails with NPE
```
Caused by: 
java.lang.NullPointerException
: 
(No message provided)
Close stacktrace
at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:890)
at org.gradle.api.internal.tasks.CachingTaskDependencyResolveContext.add(CachingTaskDependencyResolveContext.java:83)
at org.gradle.api.internal.tasks.properties.bean.AbstractNestedRuntimeBeanNode$BeanPropertyValue$2.visitDependencies(AbstractNestedRuntimeBeanNode.java:103)
at org.gradle.api.internal.tasks.CachingTaskDependencyResolveContext$TaskGraphImpl.getNodeValues(CachingTaskDependencyResolveContext.java:111)
at org.gradle.internal.graph.CachingDirectedGraphWalker$GraphWithEmptyEdges.getNodeValues(CachingDirectedGraphWalker.java:213)
at org.gradle.internal.graph.CachingDirectedGraphWalker.doSearch(CachingDirectedGraphWalker.java:121)
at org.gradle.internal.graph.CachingDirectedGraphWalker.findValues(CachingDirectedGraphWalker.java:73)
```
### Context
<!--- Which version did this stop working with? Which version do you know it works? -->
<!--- How has this issue affected you? What are you trying to accomplish? -->
<!--- Providing context helps us come up with a solution that is most useful in the real world -->

### Steps to Reproduce 
```
git clone git@github.com:eclipse/xtext-core.git
cd xtext-core
./1-gradle-build.sh
```
<!--- Provide a self-contained example project (as an attached archive or a Github project). -->
<!--- In the rare cases where this is infeasible, we will also accept a detailed set of instructions. -->

### Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->
<!--- A build scan https://scans.gradle.com/get-started is ideal -->
Build scan URL:
https://scans.gradle.com/s/btw2kfb6zf5k2/failure?openFailures=WzBd&openStackTraces=WzFd#top=0

With #9810 the NPE no longer occurs but the build still fails, this time with:

```
FAILURE: Build failed with an exception.

* What went wrong:
Circular dependency between the following tasks:
:classes
\--- :compileJava
     \--- :generateXtext
          \--- :classes (*)

(*) - details omitted (listed previously)
```

My guess is that now that properties carry task dependency information a previously hidden circular dependency might have been exposed. @cdietrich, does the new build failure ring any bells?

--- a/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/TaskInputFilePropertiesIntegrationTest.groovy
+++ b/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/TaskInputFilePropertiesIntegrationTest.groovy
@@ -150,4 +150,44 @@ class TaskInputFilePropertiesIntegrationTest extends AbstractIntegrationSpec {
         then:
         executed ":foo"
     }
+
+    @Issue("https://github.com/gradle/gradle/issues/9674")
+    def "allows @Input of task with no actions to be null"() {
+        buildFile << """
+            class FooTask extends DefaultTask {
+               @Input
+               FileCollection bar
+            }
+
+            task foo(type: FooTask)
+        """
+
+        when:
+        run "foo"
+
+        then:
+        executed ":foo"
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/9674")
+    def "shows validation error when non-Optional @Input is null"() {
+        buildFile << """
+            class FooTask extends DefaultTask {
+               @Input
+               FileCollection bar
+               
+               @TaskAction
+               def go() {
+               }
+            }
+
+            task foo(type: FooTask)
+        """
+
+        when:
+        fails "foo"
+
+        then:
+        failureCauseContains("No value has been specified for property 'bar'.")
+    }
 }

--- a/subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/bean/AbstractNestedRuntimeBeanNode.java
+++ b/subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/bean/AbstractNestedRuntimeBeanNode.java
@@ -102,7 +102,10 @@ public abstract class AbstractNestedRuntimeBeanNode extends RuntimeBeanNode<Obje
                 return new TaskDependencyContainer() {
                     @Override
                     public void visitDependencies(TaskDependencyResolveContext context) {
-                        context.add(valueSupplier.get());
+                        Object dependency = valueSupplier.get();
+                        if (dependency != null) {
+                            context.add(dependency);
+                        }
                     }
                 };
             }

